MediaRecorderPrivateEncoder can write frames out of order

jyavenard opened the pull request:
#### 4366e9326ff6f41042ca4b61d99e32d1105b1c53
<pre>
MediaRecorderPrivateEncoder can write frames out of order
<a href="https://bugs.webkit.org/show_bug.cgi?id=286805">https://bugs.webkit.org/show_bug.cgi?id=286805</a>
<a href="https://rdar.apple.com/143956063">rdar://143956063</a>

Reviewed by Youenn Fablet.

When we write to the container the video frame, we must first ensure that
we have written any earlier audio frames as WebM requires monotonically increasing timestamps.
To do so, the code could wait on in-flight audio frames prior writing the video.
However, if we paused or stopped the recorder, the waiting promise would be
immediately rejected which caused the video frame to be immediately written
only to drain/finish the audio converter which could retrieve audio frames
with a timestamp lesser than the video written.
This used to cause an assertion which was changed to a release log error instead in
287869@main.

Another issue was that it was assume we always had an audio frame with the same
presentation time as the latest video frame (as the audio clock drives the time)
However, if two video frames arrive in a burst before new audio is received
and to avoid having the video frames with the same timestamp we add 1us.
This could cause the wait for pending audio to never ends as this amended.

We can now remove some of the debug assertions we added in 287869@main and
re-introduce the ASSERT.

Added a test that exercising the code by creating a MediaRecorder with a short
timeslice.
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-AV-audio-video-dataavailable.html: Make oscillator variable global to get around webkit.org/b/271227
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-webm-AV-audio-video-dataavailable-expected.txt: Added.
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-webm-AV-audio-video-dataavailable.html: Added.
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
(WebCore::MediaRecorderPrivateEncoder::pause):
(WebCore::MediaRecorderPrivateEncoder::appendVideoFrame):
(WebCore::MediaRecorderPrivateEncoder::enqueueCompressedVideoFrame):
(WebCore::MediaRecorderPrivateEncoder::waitForMatchingAudio):
(WebCore::MediaRecorderPrivateEncoder::interleaveAndEnqueueNextFrame):
(WebCore::MediaRecorderPrivateEncoder::stopRecording):
(WebCore::MediaRecorderPrivateEncoder::flushPendingData):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.h:

Canonical link: <a href="https://commits.webkit.org/289643@main">https://commits.webkit.org/289643@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/860245d966cbea1e7ec367d3e94baad92adbb157

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87621 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7136 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42002 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/92486 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/38365 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89672 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7517 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15306 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/92486 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/38365 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90623 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79284 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/92486 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33681 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/37478 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34545 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94372 "Built successfully") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/14789 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/10842 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/76512 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/15043 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75136 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/75735 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18619 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/18536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/7740 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14805 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/20106 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/14549 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/17993 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/16331 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
```diff
@@ -212,14 +212,20 @@ void MediaRecorderPrivateEncoder::pause()
+                converter->drain()->whenSettled(queueSingleton(), [weakThis] {
+                    if (RefPtr protectedThis = weakThis.get()) {
+                        assertIsCurrent(queueSingleton());
+                        if (protectedThis->m_pendingAudioFramePromise) {
```
if (auto promise = std::exchange(protectedThis->m_pendingAudioFramePromise), { }) {
     promise->second.reject();
     LOG(MediaStream, "MediaRecorderPrivateEncoder::pause rejecting m_pendingAudioFramePromise");
}

youennf commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
```diff
@@ -212,14 +212,20 @@ void MediaRecorderPrivateEncoder::pause()
-
+            if (RefPtr converter = audioConverter()) {
+                converter->drain()->whenSettled(queueSingleton(), [weakThis] {
+                    if (RefPtr protectedThis = weakThis.get()) {
```
We tend to prefer return early for null cases like this.

youennf commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
```diff
@@ -861,20 +864,25 @@ void MediaRecorderPrivateEncoder::stopRecording()
+            converter->finish()->whenSettled(queueSingleton(), [protectedThis = Ref { *this }, this] {
+                assertIsCurrent(queueSingleton());
+
+                if (m_pendingAudioFramePromise) {
```
Ditto for std::exchange?

jyavenard committed:
MediaRecorderPrivateEncoder can write frames out of order
https://bugs.webkit.org/show_bug.cgi?id=286805
rdar://143956063

Reviewed by Youenn Fablet.

When we write to the container the video frame, we must first ensure that
we have written any earlier audio frames as WebM requires monotonically increasing timestamps.
To do so, the code could wait on in-flight audio frames prior writing the video.
However, if we paused or stopped the recorder, the waiting promise would be
immediately rejected which caused the video frame to be immediately written
only to drain/finish the audio converter which could retrieve audio frames
with a timestamp lesser than the video written.
This used to cause an assertion which was changed to a release log error instead in
287869@main.

Another issue was that it was assume we always had an audio frame with the same
presentation time as the latest video frame (as the audio clock drives the time)
However, if two video frames arrive in a burst before new audio is received
and to avoid having the video frames with the same timestamp we add 1us.
This could cause the wait for pending audio to never ends as this amended.

We can now remove some of the debug assertions we added in 287869@main and
re-introduce the ASSERT.

Added a test that exercising the code by creating a MediaRecorder with a short
timeslice.
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-AV-audio-video-dataavailable.html: Make oscillator variable global to get around webkit.org/b/271227
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-webm-AV-audio-video-dataavailable-expected.txt: Added.
* LayoutTests/http/wpt/mediarecorder/MediaRecorder-webm-AV-audio-video-dataavailable.html: Added.
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
(WebCore::MediaRecorderPrivateEncoder::pause):
(WebCore::MediaRecorderPrivateEncoder::appendVideoFrame):
(WebCore::MediaRecorderPrivateEncoder::enqueueCompressedVideoFrame):
(WebCore::MediaRecorderPrivateEncoder::waitForMatchingAudio):
(WebCore::MediaRecorderPrivateEncoder::interleaveAndEnqueueNextFrame):
(WebCore::MediaRecorderPrivateEncoder::stopRecording):
(WebCore::MediaRecorderPrivateEncoder::flushPendingData):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.h:

Canonical link: https://commits.webkit.org/289643@main
